src/test/ui/destructuring-assignment/tuple_destructure.rs

```diff
@@ -0,0 +1,24 @@
+// run-pass
+
+#![feature(destructuring_assignment)]
+
+fn main() {
+    let (mut a, mut b);
+    (a, b) = (0, 1);
+    assert_eq!((a, b), (0, 1));
+    (b, a) = (a, b);
+    assert_eq!((a, b), (1, 0));
+    (a, .., b) = (1, 2);
+    assert_eq!((a, b), (1, 2));
+    (.., a) = (1, 2);
+    assert_eq!((a, b), (2, 2));
+    (..) = (3, 4);
+    assert_eq!((a, b), (2, 2));
+    (b, ..) = (5, 6, 7);
+    assert_eq!(b, 5);
+    // Test for a non-Copy type (String):
+    let (c,d);
+    (c, d) = ("c".to_owned(), "d".to_owned());
+    assert_eq!(c, "c");
+    assert_eq!(d, "d");
```

Comment:
Could you add test cases for nested destructuring (`((a, b), c) = ...`), including nested `..` (`((a, b, ..), c) = ...`)?

Reply:
And especially `(c, d) = (d, c);` for mutable `c, d`, which should work (I manually desugared in an experiment and that worked after all)

Reply:
And some more cases:
- Parentheses shielding a tuple from being interpreted as a destructuring assignment (`((a, b)) = ...`), including a nested case (`(((a, b)), c) = ...`).
- Single-identifier constants on the left side (`(C1, C2) = ...`).

Reply:
I think this "shielding" is a bug in the current implementation. From the RFC:

> We also allow arbitrary parenthesisation, as with patterns, although unnecessary parentheses will trigger the unused_parens lint.

Reply:
Tests added as requested.